src/app.js

```diff
@@ -275,12 +276,50 @@ export default function (app) {
 
       updateElementData(element, node.data, oldNode.data)
 
-      var len = node.children.length, oldLen = oldNode.children.length
+      //All old keyed children and corresponding elements should be
+      //taken out and stored separately
+      var oldChildren = []
+      var keyedChildren = {}
+      for (var i = 0; i < oldNode.children.length; i++) {
+        var child = oldNode.children[i]
+        if (child.data && child.data._key) {
+          keyedChildren[child.data._key] = {
+            oldNode: child,
+            element: element.childNodes[i]
+          }
+        } else {
+          oldChildren.push(child)
+        }
+      }
+      for (key in keyedChildren) {
+        element.removeChild(keyedChildren[key].element)
+      }
 
-      for (var i = 0; i < len || i < oldLen; i++) {
+      var len = node.children.length, oldLen = oldChildren.length
+      for (var i = 0, j = 0; i < len || j < oldLen; i++) {
         var child = node.children[i]
+        //in case the next new node is keyed (and a previous node is known
+        //we want the element inserted at this index, and then patched
+        //normally, with the corresponding oldNode (which may be moved)
+
+        if (child && child.data && child.data._key && keyedChildren[child.data._key]) {
+          var keyed = keyedChildren[child.data._key]
+          element.appendChild(keyed.element)
```

Comment:
Probably more effincient in most cases, if this line were: `element.insertBefore(keyed.element, element.childNodes[i])`